ocd-2679.sql

```diff
@@ -0,0 +1,119 @@
+  DROP TABLE IF EXISTS openchpl.user_certification_body_map;
+  CREATE TABLE IF NOT EXISTS openchpl.user_certification_body_map (
+	id bigserial NOT NULL,
+	user_id bigint NOT NULL,
+	certification_body_id bigint NOT NULL,
+	retired bool NOT NULL DEFAULT false,
+	creation_date timestamp NOT NULL DEFAULT NOW(),
+	last_modified_date timestamp NOT NULL DEFAULT NOW(),
+	last_modified_user bigint NOT NULL,
+	deleted bool NOT NULL DEFAULT false,
+	CONSTRAINT user_certification_body_pk PRIMARY KEY (id),
+	CONSTRAINT user_fk FOREIGN KEY (user_id)
+		REFERENCES openchpl.user (user_id) 
+		MATCH SIMPLE ON UPDATE NO ACTION ON DELETE RESTRICT,
+	CONSTRAINT certification_body_fk FOREIGN KEY (certification_body_id)
+		REFERENCES openchpl.certification_body (certification_body_id) 
+		MATCH SIMPLE ON UPDATE NO ACTION ON DELETE RESTRICT
+  );
+
+  -- Populate the new table based on the ACL tables
+  INSERT INTO openchpl.user_certification_body_map
+  (user_id, certification_body_id, last_modified_user)
+  select sid.id, oi.object_id_identity, -2
+  from openchpl.acl_sid sid
+	inner join openchpl.acl_entry entry
+		on sid.id = entry.sid
+	inner join openchpl.acl_object_identity oi
+		on entry.acl_object_identity = oi.id
+	inner join openchpl.acl_class c
+		on oi.object_id_class = c.id
+  where c.class = 'gov.healthit.chpl.dto.CertificationBodyDTO'
+  and exists (select * from openchpl.user where user_id = sid.id);
+
+  --Add audit triggers
+  CREATE TRIGGER user_certification_body_map_audit AFTER INSERT OR UPDATE OR DELETE on openchpl.user_certification_body_map FOR EACH ROW EXECUTE PROCEDURE audit.if_modified_func();
+  CREATE TRIGGER user_certification_body_map_timestamp BEFORE UPDATE on openchpl.user_certification_body_map FOR EACH ROW EXECUTE PROCEDURE openchpl.update_last_modified_date_column();
+
+  -- Backup acl_entry table
+  DROP TABLE IF EXISTS openchpl.acl_entry_backup;
+
+  CREATE TABLE openchpl.acl_entry_backup AS
+  SELECT * FROM openchpl.acl_entry;
+
+  DELETE FROM openchpl.acl_entry
+  WHERE acl_object_identity IN 
+	(SELECT aoi."id" 
+	FROM openchpl.acl_object_identity aoi
+		INNER JOIN openchpl.acl_class ac
+			ON aoi.object_id_class = ac."id"
+	WHERE ac."class" = 'gov.healthit.chpl.dto.CertificationBodyDTO');
+  
+  -- Backup acl_object_identity table
+  DROP TABLE IF EXISTS openchpl.acl_object_identity_backup;
+
+  CREATE TABLE openchpl.acl_object_identity_backup AS
+  SELECT * FROM openchpl.acl_object_identity;
+
+  DELETE FROM openchpl.acl_object_identity
+  WHERE object_id_class IN 
+	(SELECT "id" 
+	 FROM openchpl.acl_class 
+	 WHERE "class" = 'gov.healthit.chpl.dto.CertificationBodyDTO');
+
+  -- Backup acl_class table
+  DROP TABLE IF EXISTS openchpl.acl_class_backup;
+
+  CREATE TABLE openchpl.acl_class_backup AS
+  SELECT * FROM openchpl.acl_class;
+
+  DELETE FROM openchpl.acl_class
+  WHERE "class" = 'gov.healthit.chpl.dto.CertificationBodyDTO';
+
+END $$;
```

Comment:
Is this supposed to be here? Does it give an error when you run the script?

Reply:
Yes, it is supposed to be there. It is the closing for 
```
DO $$
BEGIN
```
I have run the script several times, without issues.

Reply:
Why is it here? We don't usually have our SQL stuff bracketed with this

Reply:
It is necessary to the `IF ... THEN` to run correctly.  I tried removing it, and I get the following error:
```
ERROR:  syntax error at or near "IF"
LINE 3:   IF EXISTS(SELECT * FROM information_schema.tables WHERE ta...
          ^
SQL state: 42601
Character: 116
```

Reply:
My understanding of the DO blocks is that they're used for anonymous functions, where things like the IF can be used. I think I'd prefer that those anonymous functions be scoped to as small a set of code as possible. When the function has all of the other SQL statements, do those product output when you run the script? If not that's definitely an issue

Reply:
I'll wrap each IF in a DO block, for scoping purposes.  I wasn't really looking at the output -- when there is an error inside a DO, it outputs the error to the console.  When it completes successfully, there is no usable output.

Reply:
I'd be okay with all of the IFs in a single block, but I do think we need to see the other SQL update outputs as they run

Reply:
Thanks!